base/shell/explorer/notifyiconscust.cpp

```diff
@@ -20,31 +21,248 @@
+    lvColumn.cx = 190;
+    ListView_InsertColumn(hListView, 0, &lvColumn);
+
+    LoadStringW(NULL, IDS_NOTIF_BEH_BEHAVIOR, sName, _countof(sName));
+    lvColumn.pszText = sName;
+    lvColumn.cx = 130;
+    ListView_InsertColumn(hListView, 1, &lvColumn);
+
+    ListView_EnableGroupView(hListView, TRUE);
+
+    LVGROUP lvGroup;
+    lvGroup.cbSize    = sizeof(LVGROUP);
+    lvGroup.mask      = LVGF_HEADER | LVGF_ALIGN | LVGF_GROUPID | LVGF_STATE;
+    lvGroup.uAlign    = LVGA_FOOTER_LEFT;
+    lvGroup.state     = LVGS_NORMAL;
+
+    LoadStringW(NULL, IDS_NOTIF_BEH_CURRENT, sName, _countof(sName));
+    lvGroup.pszHeader = sName;
+    lvGroup.iGroupId  = 0;
+    ListView_InsertGroup(hListView, -1, &lvGroup);
+
+    LoadStringW(NULL, IDS_NOTIF_BEH_PAST, sName, _countof(sName));
+    lvGroup.pszHeader = sName;
+    lvGroup.iGroupId  = 1;
+    ListView_InsertGroup(hListView, -1, &lvGroup);
+}
+
 INT_PTR CALLBACK CustomizeNotifyIconsProc(HWND hwnd, UINT Message, WPARAM wParam, LPARAM lParam)
 {
     switch(Message)
     {
-        case WM_INITDIALOG:
-        
-        return TRUE;
+        case WM_INITDIALOG: 
+            InitializeListView(hwnd);
+            break;
+        case WM_NOTIFY:
+            {
+                if (LOWORD(wParam) == IDC_NOTIFICATION_LIST && ((LPNMHDR)lParam)->code == NM_CLICK) 
+                {
+                    ShowBehaviorCombo(hwnd, (LPNMITEMACTIVATE)lParam);
+                }
+                else return FALSE;
+            }
+            break;
         case WM_COMMAND:
-           switch(LOWORD(wParam))
-           {
-               case IDOK:
-                   EndDialog(hwnd, IDOK);
-               break;
-               case IDCANCEL:
-                   EndDialog(hwnd, IDCANCEL);
-               break;
-           }
-       break;
-       default:
-           return FALSE;
+            switch(LOWORD(wParam))
+            {
+                case IDOK:
+                    EndDialog(hwnd, IDOK);
+                    break;
+                case IDCANCEL:
+                    EndDialog(hwnd, IDCANCEL);
+                    break;
+                case IDC_NOTIFICATION_BEHAVIOUR:
+                    {
+                        if (HIWORD(wParam) == CBN_SELCHANGE)
+                        {
+                            SetItemBehavior(hwnd, (HWND)lParam);
+                        }
+                    }
+                    break;
+                case IDC_TASKBARPROP_NOTIREST:
+                    ResetBehaviors(hwnd);
+                    break;
+            }
+            break;
+        default:
+            return FALSE;
     }
     return TRUE;
 }
 
-VOID ShowCustomizeNotifyIcons(HINSTANCE hInst, HWND hExplorer)
+VOID SetNotifyIcons(HWND hDialog, HWND hTrayNotify)
+{
+    HWND hListView = GetDlgItem(hDialog, IDC_NOTIFICATION_LIST);
+    HWND hSysPager = FindWindowEx(hTrayNotify, NULL, L"SysPager", NULL);
+    HWND hToolbar  = FindWindowEx(hSysPager, NULL, L"ToolbarWindow32", NULL);
```

Comment:
Isn't there a way to retrieve the correct sub-window directly from the classes? Cc @gigaherz .

Reply:
this is a standard win32 window classes, not a custom classes with unique name. It's lucky that hTrayNotify has only one sysPager and pager has only one toolbar :)

Reply:
I was asking this because these HWNDs are just members of the corresponding CSysPager... classes, so I was wondering whether you could somehow do a pSysPager->m_hWnd somehow (with the correct names to be determined...) and/or use a correct getter function for this...

Reply:
Yeah, not a perfect way.  Tried to get toolbar HWND from CSysPager with functions like CTrayNotifyWnd_GetToolbar(m_TrayNotifyInstance) -> CSysPagerWnd_GetToolbar(&m_pager) -> Toolbar.m_hWnd field give me not null handle but it has very weird behavior:  TB_BUTTONCOUNT and Toolbar.GetButtonCount() return zero, but CNotifyToolbar.GetVisibleButtonCount() return nonzero value.

Reply:
This is definitely an ugly way to implement an obscure feature...